ocaml/lambda/lambda.mli

```diff
@@ -332,6 +332,39 @@ type primitive =
      if the value is locally allocated *)
   (* Fetching domain-local state *)
   | Pdls_get
+    (* In-place reuse of heap blocks.
+     Each of these takes the allocation to be reused as first argument.
+     The following arguments correspond to the fields that are Reuse_set. *)
+  | Preuseblock of
+      { tag : int;             (* The immutable tag of both the old and resulting block *)
+        mut : mutable_flag;                            (* Mutability of resulting block *)
+        shape : block_shape;                          (* Block shape of resulting block *)
+        resets : reset_field list;                  (* Reset-information for each field *)
+        (* Note: even if the mode is Alloc_local, the old allocation may be on the heap *)
+        mode : alloc_mode;                             (* Alloc mode of resulting block *)
+      }
+  | Preusefloatblock of
+      { mut : mutable_flag;                            (* Mutability of resulting block *)
+        resets : reset_field list;                  (* Reset-information for each field *)
+        mode : alloc_mode               (* Upper bound on alloc mode of resulting block *)
+      }
+  | Preuseufloatblock of
+      { mut : mutable_flag;                            (* Mutability of resulting block *)
+        resets : reset_field list;                  (* Reset-information for each field *)
+        mode : alloc_mode               (* Upper bound on alloc mode of resulting block *)
+      }
+  | Preusemixedblock of
+      { tag : int;             (* The immutable tag of both the old and resulting block *)
+        mut : mutable_flag;                            (* Mutability of resulting block *)
+        shape : mixed_block_shape;                    (* Block shape of resulting block *)
+        resets : reset_field list;                  (* Reset-information for each field *)
+        (* Note: even if the mode is Alloc_local, the old allocation may be on the heap *)
+        mode : alloc_mode;                             (* Alloc mode of resulting block *)
+      }
+
+and reset_field =
```

Comment:
As @anfelor and I discussed over Slack, I think it would be a significant win to change the representation here, including `reset_field` elements only when we are actually changing a field. (One way to understand this is the observation that `'a option list` is isomorphic to `(int * 'a) list`, ignoring trailing `None`s.) Not sure what the consumer of this is going to look like, so maybe we can do something even better than an `int` to count the missing fields, but storing an `int` is better than making a list with likely a lot of `None`s in it.

Reply:
See `bytegen.ml` for a consumer of this representation

Reply:
Note that each  `None` corresponds to an underscore in the source program (eg. `overwrite x with (_,3)`). We should thus expect the number of `None`s to be somewhat small and should focus just on what the best representation is for code quality.

Reply:
Which list representation do you mean? The `'a option list` in the PR or the `(int * 'a) list` proposed by Richard?

Reply:
And I don't think the number of `None`s will be all that small, in the case of record updates: `{x with f}` can have an arbitrarily large number of `None`s.

Reply:
Since Mark seemed happy with the current design, I would propose we keep the representation as it is now.